turte_graphics/turtle.js

```diff
@@ -0,0 +1,111 @@
+class Turtle {
+  constructor(x, y) {
+    this.x = x;
+    this.y = y;
+    this.position = [x, y];
+    this.path = [[x, y]];
+    this.direction = "x";
+  }
+
+  forward(steps) {
+    for (let i = 0; i < steps; i++) {
+      if (this.direction === "x") {
+        this.x++;
+      } else if (this.direction === "-y") {
+          this.y--;
+      } else if (this.direction === "-x") {
+        this.x--;
+      } else {
+          this.y++;
+      }
+      this.position = [this.x, this.y];
+      this.path.push(this.position);
+    }
+    return this;
+  }
+
+  right() {
+    if (this.direction === "x") {
+      this.direction = "-y";
+    } else if (this.direction === "-y") {
+      this.direction = "-x";
+    } else if (this.direction === "-x") {
+      this.direction = "y";
+    } else {
+      this.direction = "x";
+    }
+    return this;
+  }
+
+  left() {
+    if (this.direction === "x") {
+      this.direction = "y";
+    } else if (this.direction === "-y") {
+      this.direction = "x";
+    } else if (this.direction === "-x") {
+      this.direction = "-y";
+    } else {
+      this.direction = "-x";
+    }
+    return this;
+  }
+
+  print() {
+    const full = '\u2588';
+    const empty = " ";
+    let maxY = -Infinity;
+    let minY = Infinity;
+    let maxX = -Infinity;
+    let minX = Infinity;
+  
+    // find the highest coordinate on Y axis to begin drawing
+    for (let position of this.path) {
+      if (position[1] > maxY) {
+        maxY = position[1];
+      }
+    }
+    //find the lowest coordinate on Y axis to end drawing
+    for (let position of this.path) {
+      if (position[1] < minY) {
+        minY = position[1];
+      }
+    }
+    //find the width of the drawing with maxX and minX
+    for (let position of this.path) {
+      if (position[0] > maxX) {
+        maxX = position[0];
+      }
+    }
+    for (let position of this.path) {
+      if (position[0] < minX) {
+        minX = position[0];
+      }
+    }
+
+    console.log(`Drawing top-left: (${maxY}, ${minX}), bottom-right: (${maxX}, ${minY})`)
+
+    // draw each line
+    let drawing = "";
+    for (let y = maxY; y >= minY; y--) { // 0 --> -4
+      let row = "";
+      for (let x = minX; x <= maxX; x++) { // 0 --> 6
+        const match = this.path.find(points => {
+          return points[0] == x && points[1] == y;
+        });
```

Comment:
This loop is currently going over _all_ points in the path every time, so it's not very efficient. That said, the goal of this exercise was not efficiency and you perhaps don't need to worry about this at this point (it's called premature optimization when you address performance problems before they actually exist, which is often wasted effort). In this case, though, the change to make it more efficient is so simple that I think I would go ahead and do it anyways. 

Right now, `match` is holding _all_ the points that match the condition in the function. But that's always going to be exactly one, so you can stop looking as soon as you find a match, which you can accomplish with [`Array.findIndex`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex)

Reply:
I tried using that at first but I need it to keep looking for the rest of the coordinates where Turtle stopped at so they can all be printed not just the first one.